### ISSUE TITLE
Cannot set up the connection on spark 2.0.0

### ISSUE THREAD

**Issue by JackHo327** [rstudio/sparklyr#229]
Hi,
I cannot set up the spark_connection on spark 2.0.0, following the instruction on `http://spark.rstudio.com/deployment.html`
However, I could do that on spark 1.6.x .
For different version of spark, I've changed my SPARK_HOME.
Also, I add System32 into my sys environment variables list.

For spark 2.0.0:

> Sys.getenv("SPARK_HOME")
> [1] "D:\Spark2"
> library(sparklyr)
> sc <- spark_connect(master = "local",config = list())
> Error in start_shell(master = master, spark_home = spark_home, spark_version = version,  : 
>   Failed to launch Spark shell. Ports file does not exist.
>     Path: D:\Spark2\bin\spark-submit.cmd
>     Parameters: --class, sparklyr.Backend, "D:\R\R-3.3.1\library\sparklyr\java\sparklyr-2.0-2.11.jar", C:\Users\RENZHI~1\AppData\Local\Temp\RtmpqW0fvG\file2a84ff52ad4.out
>     The filename, directory name, or volume label syntax is incorrect.

For spark 1.6.2:

> Sys.setenv("SPARK_HOME" = "d:\Spark")
> library(sparklyr)
> sc <- spark_connect(master = "local",config = list())
> Warning message:
> In create_hive_context_v1(sc) :
>   Failed to create Hive context, falling back to SQL. Some operations, like window-functions, will not work
> library(dplyr)
> iris_tbl <- copy_to(sc, iris)
> The following columns have been renamed:
> - "Sepal.Length" => "Sepal_Length" (#1)
> - "Sepal.Width"  => "Sepal_Width"  (#2)
> - "Petal.Length" => "Petal_Length" (#3)
> - "Petal.Width"  => "Petal_Width"  (#4)
>   head(iris_tbl,5)
>   Source:   query [?? x 5]
>   Database: spark connection master=local app=sparklyr local=TRUE

  Sepal_Length Sepal_Width Petal_Length Petal_Width Species
         <dbl>       <dbl>        <dbl>       <dbl>   <chr>
1          5.1         3.5          1.4         0.2  setosa
2          4.9         3.0          1.4         0.2  setosa
3          4.7         3.2          1.3         0.2  setosa
4          4.6         3.1          1.5         0.2  setosa
5          5.0         3.6          1.4         0.2  setosa

**Comment by javierluraschi** [comment]
@JackHo327 which version of sparklyr your are running? If it's not `0.4.1` could you please upgrade using: `devtools::install_github("rstudio/sparklyr")` and retry?

The `0.4.1` contains error reporting improvements that should help us troubleshoot this issue from the new log.

**Comment by JackHo327** [comment]
Hi @javierluraschi, I am new to spark and sparklyr.
You may found that I also post something in [another ticket](https://github.com/rstudio/sparklyr/issues/114#issuecomment-249717400) , for you convenience, we could move to here.

Here is the thing, before I shift between these two versions, as my codes showed above, I checked my 'SPARK_HOME', and it pointed to "D:\Spark2", where my spark 2.0.0 located. However, I could not set up the connection, and get the error message like:
> Error in start_shell(master = master, spark_home = spark_home, spark_version = version, : 
Failed to launch Spark shell. Ports file does not exist.
Path: D:\Spark2\bin\spark-submit.cmd
Parameters: --class, sparklyr.Backend, "D:\R\R-3.3.1\library\sparklyr\java\sparklyr-2.0-2.11.jar", C:\Users\RENZHI~1\AppData\Local\Temp\RtmpqW0fvG\file2a84ff52ad4.out
The filename, directory name, or volume label syntax is incorrect.

Then I changed 'SPARK_HOME' by using Sys.setenv("SPARK_HOME" = "d:\Spark"), and then the connection was set up (although there is still a warning message):
> Warning message:
In create_hive_context_v1(sc) :
Failed to create Hive context, falling back to SQL. Some operations, like window-functions, will not work

But I think that warning just because I did not use HIVE, and I think this is acceptable, because I did not have on my laptop HIVE.

btw, do I have to use HIVE? I thought Spark and HIVE can be used separately, if I don't need to use the HIVE SQL?

Thank you.

**Comment by javierluraschi** [comment]
@JackHo327 To help troubleshoot the Hive issue, could you try to recreate the hive context as follows and report the specific error back in this thread:

```{r}
invoke_new(
      sc,
      "org.apache.spark.sql.hive.HiveContext",
      spark_context(sc)
    )
```

In general, users should be able to instantiate Hive since it's shipped as part of hadoop and supported natively in Spark.

To troubleshoot the Spark `2.0.0` issue, I would recommend to upgrade sparlyr to the latest version using the `devtools` package as follows:

```{r}
devtools::install_github("rstudio/sparklyr")
``` 

It was until version `0.4.1` that additional logging was added to help troubleshoot `spark_connect`

**Comment by JackHo327** [comment]
@javierluraschi when I input your HIVE_ISSUE codes, I get ERROR message below:

    Error: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.apache.spark.sql.hive.client.IsolatedClientLoader.createClient(IsolatedClientLoader.scala:249)
	at org.apache.spark.sql.hive.HiveContext.metadataHive$lzycompute(HiveContext.scala:327)
	at org.apache.spark.sql.hive.HiveContext.metadataHive(HiveContext.scala:237)
	at org.apache.spark.sql.hive.HiveContext.setConf(HiveContext.scala:441)
	at org.apache.spark.sql.hive.HiveContext.defaultOverrides(HiveContext.scala:226)
	at org.apache.spark.sql.hive.HiveContext.<init>(HiveContext.scala:229)
	at org.apache.spark.sql.hive.HiveContext.<init>(HiveContext.scala:101)
	at sun.reflect.NativeC

I don't think I have manually deployed any Hadoop or HIVE modules on my machine, the only thing I did was downloading Spark and R and RStuido stuff.  
**Do you think  I should download Hadoop or HIVE?** 

For my  ' sparklyr ', I use 0.4.2 actually.
   > packageVersion("sparklyr")
      [1] ‘0.4.2’

**Comment by javierluraschi** [comment]
@JackHo327 that should be enough, from your comment, seems like you installed Spark using `spark_install()`, correct? If not, which version and distribution of Spark is installed?

Could you share `sessionInfo()` to get the OS/R/etc versions in your system?

For local development, we recommend using `spark_install()`. That command should do all the installation and configuration for you so that you don't need to download or configure anything else.

This might be a long shot, but some people have hit your callstack here: http://stackoverflow.com/questions/39088433/issues-trying-out-example-in-spark-shell with something like this in sparklyr:

```{r}
conf <- spark_config()
conf[["spark.sql.warehouse.dir"]] <- "file:///C:/tmp/spark-warehouse" # you might need to tweak this for your particular machine configuration
sc <- spark_connect(master = "local", config = conf)
```

**Comment by Gnolam** [comment]
I have the same issue as **@JackHo327**.
All working great for Spark 1.6.2 but not for 2.0.0 with same  message:
> Failed to launch Spark shell. Ports file does not exist.

Java was updated, `spark_install()` used, paths all set up.

Why the error refers to "Ports file does not exist"?

**Comment by JackHo327** [comment]
Hi @javierluraschi, thanks for your help. However, I am not quite sure about :

    conf[["spark.sql.warehouse.dir"]] <- "file:///C:/tmp/spark-warehouse"

I went to check my C disk, and there is a file called 'tmp', in this tmp file, there is a file called 'Hive', and in the 'Hive', there are two interesting files called 'Renzhi He' and 'Renzhi%20He' (this file also contains another sub-level files, but they are empty). I can show you from cmd.

 ![Image1 of cmd](https://raw.githubusercontent.com/JackHo327/RwizSpark/master/terminal.png)

And I also found a file called metastore_db in my User files.

 ![Image1 of cmd](https://raw.githubusercontent.com/JackHo327/RwizSpark/master/terminal1.png)


But they all empty actually...

so, right now, what should I do? I am not quitely understand which address should I set:  

    conf[["spark.sql.warehouse.dir"]] <- "file:///C:/tmp/Hive" or "file:///C:/tmp/Renzhi He" or"file:///C:/tmp/Renzhi%20He" or "file:///C:/Users/metastore_db"

**Comment by JackHo327** [comment]
Hi @Gnolam, and yes, I also doubt whether the 'ports' refers to sth wrong with the ports of windows.

Let's see what kind of suggestions will @javierluraschi  give to us.

**Comment by menichols74** [comment]
From previous work with Spark, I have Spark 2.0 with hadoop 2.7 installed on my Win 8 computer. I have updated the env variables and can successfully run "spark-shell" or "pyspark" from the cmd line to run a scala or pyspark program (Spark is working on windows). I installed the sparklyr using the traditional "install.packages("sparklyr") and then updated to that latest using devtools (current v  0.4.5). With both versions of sparklyr, i get the same error when trying to connect R to my local spark.

`> library(sparklyr)
> Sys.getenv("SPARK_HOME")
[1] "C:\\spark-2.0.0-bin-hadoop2.7"
> 
> sc <- spark_connect(master = "local",config = list())
Error in force(code) : 
  Failed to launch Spark shell. Ports file does not exist.
    Path: C:\spark-2.0.0-bin-hadoop2.7\bin\spark-submit.cmd
    Parameters: --class, sparklyr.Backend, "C:\Program Files\R\R-3.3.1\library\sparklyr\java\sparklyr-2.0-2.11.jar", C:\Users\mnichols\AppData\Local\Temp\RtmpkFyCUI\file1e2c587d2c72.out
    
The filename, directory name, or volume label syntax is incorrect.`

I am also using the preview version of RStudio with the "Spark enhancements".  I'm OK with R and can do a little with Spark (I really like the databricks ce), but getting the two to work together seems awesome, it is just kicking my butt right now. Any ideas on what to check? I think I've tried about everything I can find in the issues and this is as far as I have gotten.

**Comment by javierluraschi** [comment]
@JackHo327 @menichols74 @Gnolam "The filename, directory name, or volume label syntax is incorrect." and other issues have been fixed in version `0.4.9`. Could you please upgrade `sparklyr` using `devtools` and `devtools::install_github("rstudio/sparklyr")`?

**Comment by menichols74** [comment]
@javierluraschi  I upgraded to version 0.4.9 and tried starting up and connecting to spark. It seemed to connect fine, but erred out when I tried to load data into spark.

`> library(sparklyr)
> sc <- spark_connect(master = "local")
> library(dplyr)
> head(iris)
  Sepal.Length Sepal.Width Petal.Length Petal.Width Species
1          5.1         3.5          1.4         0.2  setosa
2          4.9         3.0          1.4         0.2  setosa
3          4.7         3.2          1.3         0.2  setosa
4          4.6         3.1          1.5         0.2  setosa
5          5.0         3.6          1.4         0.2  setosa
6          5.4         3.9          1.7         0.4  setosa
> iris_tbl <- copy_to(sc, iris)
Error: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
	at java.lang.reflect.Constructor.newInstance(Unknown Source)
	at org.apache.spark.sql.hive.client.IsolatedClientLoader.createClient(IsolatedClientLoader.scala:258)
	at org.apache.spark.sql.hive.HiveUtils$.newClientForMetadata(HiveUtils.scala:359)
	at org.apache.spark.sql.hive.HiveUtils$.newClientForMetadata(HiveUtils.scala:263)
	at org.apache.spark.sql.hive.HiveSharedState.metadataHive$lzycompute(HiveSharedState.scala:39)
	at org.apache.spark.sql.hive.HiveSharedState.metadataHive(HiveSharedState.scala:38)
	at org.apache.spark.sql.hive.HiveSharedState.externalCatalog$lzycompute(HiveSharedState.scala:46)
	at org.apache.spark.sql.hive.HiveSharedState.externalCatalog(HiveSharedState.scala:45)
	at org.apa
`
Again, this is on Win8, RStudio preview version with Spark, installed via "spark_install(version = "2.0.0", hadoop_version = "2.7")" run within R.

**Comment by javierluraschi** [comment]
@menichols74 Alright, so the good news, I think we are past this being a connection issue to Spark. Instead, from the callstack, this seems to be a hive configurations issue.

I, myself, have not been able to hit this in Windows. From @JackHo327 comments, sounds like using something like the following helped address this:

```{r}
config <- spark_config()
config[["spark.sql.warehouse.dir"]] <- "file:C:/temp"
spark_connect(master = "local", config = config)
```

However, I haven't been able to fully reproduce this. If I can hear enough feedback that this setting is unblocking users of `sparklyr`, we could consider enabling it by default in out configuration.

Related: https://github.com/rstudio/sparklyr/issues/189 and this JIRA issue https://issues.apache.org/jira/browse/SPARK-15565 and this PR https://github.com/apache/spark/pull/13348

**Comment by JackHo327** [comment]
Hi @menichols74 , here is my situation:

- my **R session info** listed below:
> R version 3.3.1 (2016-06-21)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

> locale:
[1] LC_COLLATE=Chinese (Simplified)_China.936 
[2] LC_CTYPE=Chinese (Simplified)_China.936   
[3] LC_MONETARY=Chinese (Simplified)_China.936
[4] LC_NUMERIC=C                              
[5] LC_TIME=Chinese (Simplified)_China.936    

> attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

> loaded via a namespace (and not attached):
[1] tools_3.3.1

- I use **RStudio Preview Version 1.0.25** – © 2009-2016 RStudio, Inc.

- my **sparklyr** version is:
> packageVersion("sparklyr")
[1] ‘0.4.9’

### If I just set up the connection without setting my `spark.sql.warehouse.dir`, things will happen like:
    > library(sparklyr)
    > library(dplyr)
    > sc <- spark_connect(master = "spark://spark01.cmua.dom:7077")
    > iris_tbl <- copy_to(sc, iris)
> Error: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.apache.spark.sql.hive.client.IsolatedClientLoader.createClient(IsolatedClientLoader.scala:258)
	at org.apache.spark.sql.hive.HiveUtils$.newClientForMetadata(HiveUtils.scala:359)
	at org.apache.spark.sql.hive.HiveUtils$.newClientForMetadata(HiveUtils.scala:263)
	at org.apache.spark.sql.hive.HiveSharedState.metadataHive$lzycompute(HiveSharedState.scala:39)
	at org.apache.spark.sql.hive.HiveSharedState.metadataHive(HiveSharedState.scala:38)
	at org.apache.spark.sql.hive.HiveSharedState.externalCatalog$lzycompute(HiveSharedState.scala:46)
	at org.apache.spark.sql.hive.HiveSharedStat

### However
if I set with my `spark.sql.warehouse.dir`, then:
    > library(sparklyr)
    > library(dplyr)
    > sc <- spark_connect(master = "spark://spark01.cmua.dom:7077", config = list(spark.sql.warehouse.dir = "d:/Spark2/tmp/hadoop/bin"))
    > iris_tbl <- copy_to(sc, iris)
The following columns have been renamed:
- "Sepal.Length" => "Sepal_Length" (#1)
- "Sepal.Width"  => "Sepal_Width"  (#2)
- "Petal.Length" => "Petal_Length" (#3)
- "Petal.Width"  => "Petal_Width"  (#4)
    > head(iris_tbl,4)
Source:   query [?? x 5]
Database: spark connection master=spark://spark01.cmua.dom:7077 app=sparklyr local=FALSE

  Sepal_Length Sepal_Width Petal_Length Petal_Width Species
         <dbl>       <dbl>        <dbl>       <dbl>   <chr>
1          5.1         3.5          1.4         0.2  setosa
2          4.9         3.0          1.4         0.2  setosa
3          4.7         3.2          1.3         0.2  setosa
4          4.6         3.1          1.5         0.2  setosa

In fact, my `spark.sql.warehouse.dir` just point to a file where has a winutils.exe, I am not very sure about what it is, seems like a Hadoop or Hive stuff.

Hope these can be a little helpful.

**Comment by javierluraschi** [comment]
@JackHo327 Thank you for the detailed explanation and results you obtain with/without this setting!

Opened this issue https://github.com/rstudio/sparklyr/issues/250 to consider integrating your setting as part of the automatic configuration.

**Comment by Gnolam** [comment]
Hi,

**@javierluraschi** thank you for issuing new dev version of the sparklyr package.
I currently have `sparklyr_0.4.11` and 

> Sys.setenv(SPARK_HOME = 'C:/spark/spark-2.0.0-bin-hadoop2.7')
> options(spark.install.dir = "c:/spark")
> sc <- spark_connect(master = "local")
> iris_tbl <- copy_to(sc, iris, "iris") 

> iris_tbl %>%
  select(Petal_Length, Petal_Width) %>%
  top_n(40, Petal_Width) %>%
  arrange(Petal_Length)

work just fine. Thank you very much

**Comment by menichols74** [comment]
@javierluraschi @JackHo327 @Gnolam  I really do appreciate the help. Prior to updating to 0.4.11, I tried Jack's suggestion and set my "config = list(... to the location of the winutils) and I still got the same error message from my Oct 4 post. I thought maybe something was off with my initial install of spark and where things were placed (I had multiple versions of Spark installed from Apache site and using RStudio and the sparklyr package). I removed all spark packages and upgraded to sparklyr 0.4.11. Then used installed a fresh version of spark using:

`> spark_install(version = "2.0.0", hadoop_version = "2.7")
Installing Spark 2.0.0 for Hadoop 2.7 or later.
Downloading from:
- 'https://d3kbcqa49mib13.cloudfront.net/spark-2.0.0-bin-hadoop2.7.tgz'
Installing to:
- 'C:\Users\mnichols\AppData\Local\rstudio\spark\Cache/spark-2.0.0-bin-hadoop2.7'
trying URL 'https://d3kbcqa49mib13.cloudfront.net/spark-2.0.0-bin-hadoop2.7.tgz'
Content type 'application/x-tar' length 186354175 bytes (177.7 MB)
downloaded 177.7 MB

Installation complete.
`
To verify that spark was installed properly I ran >spark-shell from a command line and Spark started up just fine. I quit spark and exited the cmd window. Then in R I tried both @JackHo327  and @Gnolam ways with the same result, I'm now no longer even getting the spark connection established:

`
> Sys.setenv(SPARK_HOME = 'C:/Users/mnichols/AppData/Local/rstudio/spark/Cache/spark-2.0.0-bin-hadoop2.7')
> options(spark.install.dir = "C:/Users/mnichols/AppData/Local/rstudio/spark")
> sc <- spark_connect(master = "local")
Created default hadoop bin directory under: C:\Users\mnichols\AppData\Local\rstudio\spark\Cache\spark-2.0.0-bin-hadoop2.7\tmp\hadoop
Error in force(code) : 
  Failed while connecting to sparklyr to port (8880) for sessionid (4085): Gateway in port (8880) did not respond.
    Path: C:\Users\mnichols\AppData\Local\rstudio\spark\Cache\spark-2.0.0-bin-hadoop2.7\bin\spark-submit.cmd
    Parameters: --class, sparklyr.Backend, --packages, com.databricks:spark-csv_2.11:1.3.0, C:\Users\mnichols\Documents\R\win-library\3.3\sparklyr\java\sparklyr-2.0-2.11.jar, 8880, 4085
    Traceback:
      shell_connection(master = master, spark_home = spark_home, app_name = app_name, version = version, hadoop_version = hadoop_version, shell_args = shell_args, config = config, service = FALSE, extensions = extensions)
      start_shell(master = master, spark_home = spark_home, spark_version = version, app_name = app_name, config = config, jars = spark_config_value(config, "spark.jars.default", list()), packages = spark_config_value(config, "sparklyr.defaultPackages"), extensions = extensions, environment = environment, shell_args = shell_args, service = service)
      tryCatch({
    gatewayInfo <- spark_connect_gateway(gatewayAddress, gatewayPort, sessionId, config = config, isStarting = TRUE)
}, error = function(e) {
    abort_shell(paste("Failed while connecting to sparklyr to port (", gatewayPort, ") for sessionid (", sessionId, "): ", e$message, sep = ""), spark_submit_path, shell_args, output_file, error_file)
})
      tryCatchList(expr, classes, parentenv, handlers)
      tryCatchOne(expr, names, parentenv, handlers[[1]])
      value[[3]](cond)
      abort_shell(paste("Failed while connecting to sparklyr to port (", gatewayPort, ") for sessionid (", sessionId, "): ", e$message, sep = ""), spark_submit_path, shell_args, output_file, error_file)

---- Output Log ----


---- Error Log ----
In addition: Warning message:
running command '"C:\Users\mnichols\AppData\Local\rstudio\spark\Cache\spark-2.0.0-bin-hadoop2.7\bin\spark-submit.cmd" --class sparklyr.Backend --packages com.databricks:spark-csv_2.11:1.3.0 C:\Users\mnichols\Documents\R\win-library\3.3\sparklyr\java\sparklyr-2.0-2.11.jar 8880 4085' had status 127 
> `

I think this stays in this issue, trouble connecting to Spark.

**Comment by javierluraschi** [comment]
@menichols74 one more update, please try with `sparklyr` `0.4.13`. Thanks.

**Comment by menichols74** [comment]
@javierluraschi  I'm certain it is user error on my part, but I have tried it now with Apache and Sparklyr versions of installed spark (both of which I can run from CMD). I've got back to establishing a connection, but the same Java errors when I try to load data to the spark context that @JackHo327 addressed earlier. I printed out the sc list in this instance, maybe there is a clue to what I'm doing wrong.

`> library(sparklyr)
> options(spark.install.dir = "C:/spark/spark-2.0.1-bin-hadoop2.7")
> sc <- spark_connect(master = "local", config = list(spark.sql.warehouse.dir = "C:/spark/spark-2.0.1-bin-hadoop2.7/tmp/hadoop/bin"))
> library(dplyr)
> iris_tbl <- copy_to(sc, iris)
Error: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
	at java.lang.reflect.Constructor.newInstance(Unknown Source)
	at org.apache.spark.sql.hive.client.IsolatedClientLoader.createClient(IsolatedClientLoader.scala:258)
	at org.apache.spark.sql.hive.HiveUtils$.newClientForMetadata(HiveUtils.scala:359)
	at org.apache.spark.sql.hive.HiveUtils$.newClientForMetadata(HiveUtils.scala:263)
	at org.apache.spark.sql.hive.HiveSharedState.metadataHive$lzycompute(HiveSharedState.scala:39)
	at org.apache.spark.sql.hive.HiveSharedState.metadataHive(HiveSharedState.scala:38)
	at org.apache.spark.sql.hive.HiveSharedState.externalCatalog$lzycompute(HiveSharedState.scala:46)
	at org.apache.spark.sql.hive.HiveSharedState.externalCatalog(HiveSharedState.scala:45)
	at org.apa
> print(sc)
$master
[1] "local"

$method
[1] "shell"

$app_name
[1] "sparklyr"

$config
$config$spark.sql.warehouse.dir
[1] "C:/spark/spark-2.0.1-bin-hadoop2.7/tmp/hadoop/bin"


$spark_home
[1] "C:\\spark\\spark-2.0.1-bin-hadoop2.7"

$backend
        description               class                mode                text              opened            can read 
"->localhost:64277"          "sockconn"                "wb"            "binary"            "opened"               "yes" 
          can write 
              "yes" 

$monitor
       description              class               mode               text             opened           can read          can write 
"->localhost:8880"         "sockconn"               "rb"           "binary"           "opened"              "yes"              "yes" 

$output_file
[1] "C:\\Users\\mnichols\\AppData\\Local\\Temp\\Rtmpyed5b2\\filef64170435b3_spark.log"

$spark_context
<jobj[4]>
  class org.apache.spark.SparkContext
  org.apache.spark.SparkContext@1697f7f

$java_context
<jobj[5]>
  class org.apache.spark.api.java.JavaSparkContext
  org.apache.spark.api.java.JavaSparkContext@1d62d32

$hive_context
<jobj[8]>
  class org.apache.spark.sql.SparkSession
  org.apache.spark.sql.SparkSession@16e6762

attr(,"class")
[1] "spark_connection"       "spark_shell_connection" "DBIConnection"         
> 
`

**Comment by javierluraschi** [comment]
@menichols74 See https://github.com/rstudio/sparklyr/issues/189. Could you try using this:

```{r}
config <- spark_config()
hadoopBin <- normalizePath(file.path(spark_home_dir(), "tmp", "hadoop", "bin"))
config[["spark.sql.warehouse.dir"]] <- if (.Platform$OS.type == "windows") hadoopBin else NULL
sc <- spark_connect(master = "local", config = config)
```

Let me know if this works for you since this fix is generic enough that I'm comfortable pushing it out by default in `sparklyr`.

**Comment by upton9265** [comment]
Hi All,

I believed I followed the directions above, but still getting the java reflection error. Below is my sequence of commands, and R sessionInfo (running from the plain R console), and a dump of my session context:
devtools::install_github("rstudio/sparklyr")
library(sparklyr)
spark_install(version="2.0.0")

library(dplyr)

config <- spark_config()
hadoopBin <- normalizePath(file.path(spark_home_dir(), "tmp", "hadoop", "bin"))
config[["spark.sql.warehouse.dir"]] <- if (.Platform$OS.type == "windows") hadoopBin else NULL
sc <- spark_connect(master = "local", config = config)

<<<  at this point, I got a "Unsupported 16-Bit application" dialog popping up on windows, so this may be the main issue for me. I didn't see anywhere at https://github.com/steveloughran/winutils where there was a distinction between 64-bit and 32-bit winutils.exe version, so I may have missed something there. I downloaded winutils.exe from here: https://github.com/steveloughran/winutils/tree/master/hadoop-2.7.1/bin>>>

flights_tbl <- copy_to(sc,nycflights13::flights,"flights")

Error: java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at org.apache.spark.sql.hive.client.IsolatedClientLoader.createClient(IsolatedClientLoader.scala:258)
        at org.apache.spark.sql.hive.HiveUtils$.newClientForMetadata(HiveUtils.scala:359)
        at org.apache.spark.sql.hive.HiveUtils$.newClientForMetadata(HiveUtils.scala:263)
        at org.apache.spark.sql.hive.HiveSharedState.metadataHive$lzycompute(HiveSharedState.scala:39)
        at org.apache.spark.sql.hive.HiveSharedState.metadataHive(HiveSharedState.scala:38)
        at org.apache.spark.sql.hive.HiveSharedState.externalCatalog$lzycompute(HiveSharedState.scala:46)
        at org.apache.spark.sql.hive.HiveSharedStat

		
R version 3.3.1 (2016-06-21)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 7 x64 (build 7601) Service Pack 1

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] dplyr_0.5.0     sparklyr_0.4.14

loaded via a namespace (and not attached):
 [1] magrittr_1.5    R6_2.1.2        assertthat_0.1  rprojroot_1.0-2 parallel_3.3.1 
 [6] tools_3.3.1     DBI_0.5-1       withr_1.0.2     yaml_2.1.13     rappdirs_0.3.1 
[11] tibble_1.2      Rcpp_0.12.6     config_0.2      digest_0.6.10  




> sc
$master
[1] "local[2]"

$method
[1] "shell"

$app_name
[1] "sparklyr"

$config
$config$sparklyr.cores.local
[1] 2

$config$spark.sql.shuffle.partitions.local
[1] 2

$config$spark.env.SPARK_LOCAL_IP.local
[1] "127.0.0.1"

$config$sparklyr.defaultPackages
[1] "com.databricks:spark-csv_2.11:1.3.0"

$config$spark.sql.warehouse.dir
[1] "file:C:/temp"

attr(,"config")
[1] "default"
attr(,"file")
[1] "C:\\R\\R-3.3.1\\library\\sparklyr\\conf\\config-template.yml"

$spark_home
[1] "C:\\Users\\Stephen Upton\\AppData\\Local\\rstudio\\spark\\Cache\\spark-2.0.0-bin-hadoop2.7"

$backend
        description               class                mode                text 
"->localhost:51105"          "sockconn"                "wb"            "binary" 
             opened            can read           can write 
           "opened"               "yes"               "yes" 

$monitor
       description              class               mode               text             opened 
"->localhost:8880"         "sockconn"               "rb"           "binary"           "opened" 
          can read          can write 
             "yes"              "yes" 

$output_file
[1] "C:\\Users\\STEPHE~1\\AppData\\Local\\Temp\\RtmpczdvT7\\file1ae45af94b71_spark.log"

$spark_context
<jobj[5]>
  class org.apache.spark.SparkContext
  org.apache.spark.SparkContext@ec055b2

$java_context
<jobj[6]>
  class org.apache.spark.api.java.JavaSparkContext
  org.apache.spark.api.java.JavaSparkContext@7194e2c3

$hive_context
<jobj[9]>
  class org.apache.spark.sql.SparkSession
  org.apache.spark.sql.SparkSession@1da8e9cc

attr(,"class")
[1] "spark_connection"       "spark_shell_connection" "DBIConnection"

**Comment by javierluraschi** [comment]
@upton9265 the error you mention does point out to an issue using the incorrect version of `WINUTILS`. We have logic to suggest which version of WINUTILS you should download when not running from within an RStudio session, try running:

```{r}
sparklyr:::stop_with_winutils_error(hadoopBinPath = NULL)
```

and downloading `winutils` from the suggested path. This should either suggest a download from:

https://github.com/steveloughran/winutils/raw/master/hadoop-2.6.0/bin/ or
https://github.com/steveloughran/winutils/raw/master/hadoop-2.6.0/bin/

than then you would have to copy to `tmp/hadoop/bin` directory. If you delete `winutils` and attempt to run `sc <- spark_connect(master = "local")`, you should get an error message with instructions on how to proceed.

Let us know how this goes.

**Comment by upton9265** [comment]
@javierluraschi Both URLs that you list above are exactly the same. In any case, I downloaded the entire folder at https://github.com/steveloughran/winutils/raw/master/hadoop-2.6.0/bin/ and that go rid of the dialog box saying WINUTILS was bad, but I'm still getting the java reflection error, so at least some progress!!

BTW, I don't have to load rJava or any other R package, correct? I'm also able to run spark 2.0.0 from the command line from another installation of spark 2.0 at c:/spark. That shouldn't make a difference either?

**Comment by yarrick19** [comment]
I am also having similar issues getting sparklyr to successfully connect to Spark on my windows 10 machine.  Below is the console output showing the commands I ran and the results.  This was all run through the preview version of RStudio.

Does anyone have any other advice on what to try?  Is there any other information I can supply that would be helpful in tracing this issue down?

```
> devtools::install_github("rstudio/sparklyr")
Downloading GitHub repo rstudio/sparklyr@master
from URL https://api.github.com/repos/rstudio/sparklyr/zipball/master
Installing sparklyr
"C:/PROGRA~1/R/R-33~1.1/bin/x64/R" --no-site-file --no-environ --no-save --no-restore --quiet CMD INSTALL  \
  "C:/Users/pnieman/AppData/Local/Temp/RtmpkhAS5H/devtools45b41664732b/rstudio-sparklyr-e31cb67" --library="C:/Users/pnieman/Documents/R/win-library/3.3"  \
  --install-tests 

* installing *source* package 'sparklyr' ...
** R
** inst
** tests
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded
*** arch - i386
*** arch - x64
* DONE (sparklyr)
Reloading installed sparklyr

> spark_install(version = "2.0.0")
Installing Spark 2.0.0 for Hadoop 2.7 or later.
Downloading from:
- 'https://d3kbcqa49mib13.cloudfront.net/spark-2.0.0-bin-hadoop2.7.tgz'
Installing to:
- 'C:\Users\pnieman\AppData\Local\rstudio\spark\Cache/spark-2.0.0-bin-hadoop2.7'
trying URL 'https://d3kbcqa49mib13.cloudfront.net/spark-2.0.0-bin-hadoop2.7.tgz'
Content type 'application/x-tar' length 186354175 bytes (177.7 MB)
downloaded 177.7 MB

cygwin warning:
  MS-DOS style path detected: C:\Users\pnieman\AppData\Local\rstudio\spark\Cache/spark-2.0.0-bin-hadoop2.7.tgz
  Preferred POSIX equivalent is: /cygdrive/c/Users/pnieman/AppData/Local/rstudio/spark/Cache/spark-2.0.0-bin-hadoop2.7.tgz
  CYGWIN environment variable option "nodosfilewarning" turns off this warning.
  Consult the user's guide for more details about POSIX paths:
    http://cygwin.com/cygwin-ug-net/using.html#using-pathnames
Installation complete.

> library(sparklyr)
> library(dplyr)

Attaching package: ‘dplyr’

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union

> sc <- spark_connect(master = "local")
Created default hadoop bin directory under: C:\Users\pnieman\AppData\Local\rstudio\spark\Cache\spark-2.0.0-bin-hadoop2.7\tmp\hadoop
Error in force(code) : 
  Failed while connecting to sparklyr to port (8880) for sessionid (7874): Gateway in port (8880) did not respond.
    Path: C:\Users\pnieman\AppData\Local\rstudio\spark\Cache\spark-2.0.0-bin-hadoop2.7\bin\spark-submit2.cmd
    Parameters: --class, sparklyr.Backend, --packages, "com.databricks:spark-csv_2.11:1.3.0", "C:\Users\pnieman\Documents\R\win-library\3.3\sparklyr\java\sparklyr-2.0-2.11.jar", 8880, 7874
    Traceback:
      shell_connection(master = master, spark_home = spark_home, app_name = app_name, version = version, hadoop_version = hadoop_version, shell_args = shell_args, config = config, service = FALSE, extensions = extensions)
      start_shell(master = master, spark_home = spark_home, spark_version = version, app_name = app_name, config = config, jars = spark_config_value(config, "spark.jars.default", list()), packages = spark_config_value(config, "sparklyr.defaultPackages"), extensions = extensions, environment = environment, shell_args = shell_args, service = service)
      tryCatch({
    gatewayInfo <- spark_connect_gateway(gatewayAddress, gatewayPort, sessionId, config = config, isStarting = TRUE)
}, error = function(e) {
    abort_shell(paste("Failed while connecting to sparklyr to port (", gatewayPort, ") for sessionid (", sessionId, "): ", e$message, sep = ""), spark_submit_path, shell_args, output_file, error_file)
})
      tryCatchList(expr, classes, parentenv, handlers)
      tryCatchOne(expr, names, parentenv, handlers[[1]])
      value[[3]](cond)
      abort_shell(paste("Failed while connecting to sparklyr to port (", gatewayPort, ") for sessionid (", sessionId, "): ", e$message, sep = ""), spark_submit_path, shell_args, output_file, error_file)

---- Output Log ----


---- Error Log ----
In addition: Warning message:
running command '"C:\Users\pnieman\AppData\Local\rstudio\spark\Cache\spark-2.0.0-bin-hadoop2.7\bin\spark-submit2.cmd" --class sparklyr.Backend --packages "com.databricks:spark-csv_2.11:1.3.0" "C:\Users\pnieman\Documents\R\win-library\3.3\sparklyr\java\sparklyr-2.0-2.11.jar" 8880 7874' had status 127 
> 
> config <- spark_config()
> hadoopBin <- normalizePath(file.path(spark_home_dir(), "tmp", "hadoop", "bin"))
> config[["spark.sql.warehouse.dir"]] <- if (.Platform$OS.type == "windows") hadoopBin else NULL
> sc <- spark_connect(master = "local", config = config)
Error in force(code) : 
  Failed while connecting to sparklyr to port (8880) for sessionid (6321): Gateway in port (8880) did not respond.
    Path: C:\Users\pnieman\AppData\Local\rstudio\spark\Cache\spark-2.0.0-bin-hadoop2.7\bin\spark-submit2.cmd
    Parameters: --class, sparklyr.Backend, --packages, "com.databricks:spark-csv_2.11:1.3.0", "C:\Users\pnieman\Documents\R\win-library\3.3\sparklyr\java\sparklyr-2.0-2.11.jar", 8880, 6321
    Traceback:
      shell_connection(master = master, spark_home = spark_home, app_name = app_name, version = version, hadoop_version = hadoop_version, shell_args = shell_args, config = config, service = FALSE, extensions = extensions)
      start_shell(master = master, spark_home = spark_home, spark_version = version, app_name = app_name, config = config, jars = spark_config_value(config, "spark.jars.default", list()), packages = spark_config_value(config, "sparklyr.defaultPackages"), extensions = extensions, environment = environment, shell_args = shell_args, service = service)
      tryCatch({
    gatewayInfo <- spark_connect_gateway(gatewayAddress, gatewayPort, sessionId, config = config, isStarting = TRUE)
}, error = function(e) {
    abort_shell(paste("Failed while connecting to sparklyr to port (", gatewayPort, ") for sessionid (", sessionId, "): ", e$message, sep = ""), spark_submit_path, shell_args, output_file, error_file)
})
      tryCatchList(expr, classes, parentenv, handlers)
      tryCatchOne(expr, names, parentenv, handlers[[1]])
      value[[3]](cond)
      abort_shell(paste("Failed while connecting to sparklyr to port (", gatewayPort, ") for sessionid (", sessionId, "): ", e$message, sep = ""), spark_submit_path, shell_args, output_file, error_file)

---- Output Log ----


---- Error Log ----
In addition: Warning message:
running command '"C:\Users\pnieman\AppData\Local\rstudio\spark\Cache\spark-2.0.0-bin-hadoop2.7\bin\spark-submit2.cmd" --class sparklyr.Backend --packages "com.databricks:spark-csv_2.11:1.3.0" "C:\Users\pnieman\Documents\R\win-library\3.3\sparklyr\java\sparklyr-2.0-2.11.jar" 8880 6321' had status 127 
> 
> sessionInfo()
R version 3.3.1 (2016-06-21)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] dplyr_0.5.0     sparklyr_0.4.15

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.7     digest_0.6.10   withr_1.0.2     rprojroot_1.0-2 assertthat_0.1  rappdirs_0.3.1  R6_2.2.0        DBI_0.5-1       git2r_0.15.0    magrittr_1.5   
[11] httr_1.2.1      curl_2.1        config_0.2      devtools_1.12.0 tools_3.3.1     parallel_3.3.1  yaml_2.1.13     memoise_1.0.0   tibble_1.2     
```

**Comment by javierluraschi** [comment]
@upton9265 no, no additional packages are needed and no, it won't make a difference to run it from a different location. Which reflection exception are we talking about? Could you share your logs. Once you are past connection issues in `sparklyr`, it usually helps to find the exception in Spark forums since you might find the solution to the answer within the Spark community. That said, share your trace and I'll take a look.

**Comment by javierluraschi** [comment]
@yarrick19 saw your post in a different github issue, I asked for more info in that issue so would be better to follow up there.

**Comment by upton9265** [comment]
@javierluraschi Here's a dump of my RGui console, ending with the Error: java.lang.reflect.InvocationTargetException. I will take your advice and take a look on the Spark forums, but any knowledge of why this might occur would be most appreciated.


> devtools::install_github("rstudio/sparklyr")
Downloading GitHub repo rstudio/sparklyr@master
from URL https://api.github.com/repos/rstudio/sparklyr/zipball/master
Installing sparklyr
"C:/R/R-3.3.1/bin/x64/R" --no-site-file --no-environ --no-save --no-restore --quiet CMD  \
  INSTALL "C:/Users/Stephen  \
  Upton/AppData/Local/Temp/RtmpgLxDay/devtools19707d742aec/rstudio-sparklyr-8d142c4"  \
  --library="C:/R/R-3.3.1/library" --install-tests 

* installing *source* package 'sparklyr' ...
** R
** inst
** tests
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded
*** arch - i386
*** arch - x64
* DONE (sparklyr)
> library(sparklyr)
> sessionInfo()
R version 3.3.1 (2016-06-21)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 7 x64 (build 7601) Service Pack 1

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] sparklyr_0.4.16

loaded via a namespace (and not attached):
 [1] httr_1.2.1      magrittr_1.5    assertthat_0.1  R6_2.1.2        rprojroot_1.0-2
 [6] DBI_0.5-1       tools_3.3.1     withr_1.0.2     dplyr_0.5.0     tibble_1.2     
[11] curl_1.1        Rcpp_0.12.6     memoise_1.0.0   git2r_0.15.0    digest_0.6.10  
[16] devtools_1.12.0
> spark_install(version="2.0.0")
Spark 2.0.0 for Hadoop 2.7 or later already installed.
> library(dplyr)

Attaching package: ‘dplyr’

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union

> Sys.getenv()
ALLUSERSPROFILE             C:\ProgramData
APPDATA                     C:\Users\Stephen Upton\AppData\Roaming
CHARSET                     cp1252
CLASSPATH                   ;C:/AtomicScala/examples
CommonProgramFiles          C:\Program Files\Common Files
CommonProgramFiles(x86)     C:\Program Files (x86)\Common Files
CommonProgramW6432          C:\Program Files\Common Files
COMPUTERNAME                WIN-FU98SDEF4P7
ComSpec                     C:\Windows\system32\cmd.exe
DEFLOGDIR                   C:\ProgramData\McAfee\DesktopProtection
FP_NO_HOST_CHECK            NO
GFORTRAN_STDERR_UNIT        -1
GFORTRAN_STDOUT_UNIT        -1
GIT_SSH                     C:\Program Files\TortoiseGit\bin\TortoiseGitPlink.exe
HADOOP_HOME                 C:\spark\spark-1.5.1-bin-hadoop2.6\tmp\hadoop
HOME                        C:\Users\Stephen Upton\Documents
HOMEDRIVE                   C:
HOMEPATH                    \Users\Stephen Upton
JAVA_HOME                   C:\Program Files\Java\jdk1.8.0_101
LDMS_LOCAL_DIR              C:\Program Files (x86)\LANDesk\LDClient\Data
LOCALAPPDATA                C:\Users\Stephen Upton\AppData\Local
LOGONSERVER                 \\WIN-FU98SDEF4P7
NUMBER_OF_PROCESSORS        2
OS                          Windows_NT
PATH                        C:\spark\spark-2.0.0-bin-hadoop2.7/bin;C:\Program
                            Files\Java\jdk1.8.0_101\bin;C:\ProgramData\Oracle\Java\javapath;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Program
                            Files\ActivIdentity\ActivClient\;C:\Program Files
                            (x86)\ActivIdentity\ActivClient\;C:\scala\2.11.8\bin;C:\Windows\system32\config\systemprofile\.dnx\bin;C:\Program
                            Files\Microsoft
                            DNX\Dnvm\;C:\R\R-3.3.1\bin\x64;C:\Program Files
                            (x86)\Windows Kits\8.1\Windows Performance
                            Toolkit\;C:\Program Files
                            (x86)\PuTTY\;C:\condor\bin\;C:\Program Files
                            (x86)\sbt\bin;C:\BaseX\bin;C:\scala\2.11.8\bin;C:\Program
                            Files\TortoiseGit\bin;C:\Program
                            Files\Git\cmd;C:\python\Anaconda3;C:\python\Anaconda3\Scripts;C:\python\Anaconda3\Library\bin;C:\Users\Stephen
                            Upton\AppData\Local\atom\bin
PATHEXT                     .COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC
PROCESSOR_ARCHITECTURE      AMD64
PROCESSOR_IDENTIFIER        Intel64 Family 6 Model 58 Stepping 9, GenuineIntel
PROCESSOR_LEVEL             6
PROCESSOR_REVISION          3a09
ProgramData                 C:\ProgramData
ProgramFiles                C:\Program Files
ProgramFiles(x86)           C:\Program Files (x86)
ProgramW6432                C:\Program Files
PSModulePath                C:\Windows\system32\WindowsPowerShell\v1.0\Modules\
PUBLIC                      C:\Users\Public
R_ARCH                      /x64
R_COMPILED_BY               gcc 4.9.3
R_HOME                      C:/R/R-3.3.1
R_LIBS_USER                 C:\Users\Stephen Upton\Documents/R/win-library/3.3
R_USER                      C:\Users\Stephen Upton\Documents
SBT_HOME                    C:\Program Files (x86)\sbt\
SCALA_HOME                  C:\scala\2.11.8
SESSIONNAME                 Console
SPARK_HOME                  C:\spark\spark-2.0.0-bin-hadoop2.7
SVN_SSH                     C:\Program Files\TortoiseGit\bin\TortoiseGitPlink.exe
SystemDrive                 C:
SystemRoot                  C:\Windows
TEMP                        C:\Users\STEPHE~1\AppData\Local\Temp
TMP                         C:\Users\STEPHE~1\AppData\Local\Temp
USERDOMAIN                  WIN-FU98SDEF4P7
USERNAME                    Stephen Upton
USERPROFILE                 C:\Users\Stephen Upton
VS140COMNTOOLS              C:\Program Files (x86)\Microsoft Visual Studio
                            14.0\Common7\Tools\
VSEDEFLOGDIR                C:\ProgramData\McAfee\DesktopProtection
windir                      C:\Windows
windows_tracing_flags       3
windows_tracing_logfile     C:\BVTBin\Tests\installpackage\csilogfile.log
> spark_home_dir()
[1] "C:\\Users\\Stephen Upton\\AppData\\Local\\rstudio\\spark\\Cache/spark-2.0.0-bin-hadoop2.7"
> config <- spark_config()
> hadoopBin <- normalizePath(file.path(spark_home_dir(), "tmp", "hadoop", "bin"))
> config[["spark.sql.warehouse.dir"]] <- if (.Platform$OS.type == "windows") hadoopBin else NULL
> sc <- spark_connect(master = "local", config = config)
> flights_tbl <- copy_to(sc,nycflights13::flights,"flights")
Error: java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at org.apache.spark.sql.hive.client.IsolatedClientLoader.createClient(IsolatedClientLoader.scala:258)
        at org.apache.spark.sql.hive.HiveUtils$.newClientForMetadata(HiveUtils.scala:359)
        at org.apache.spark.sql.hive.HiveUtils$.newClientForMetadata(HiveUtils.scala:263)
        at org.apache.spark.sql.hive.HiveSharedState.metadataHive$lzycompute(HiveSharedState.scala:39)
        at org.apache.spark.sql.hive.HiveSharedState.metadataHive(HiveSharedState.scala:38)
        at org.apache.spark.sql.hive.HiveSharedState.externalCatalog$lzycompute(HiveSharedState.scala:46)
        at org.apache.spark.sql.hive.HiveSharedStat
>

**Comment by upton9265** [comment]
@javierluraschi I may have stumbled across something related to the above Exception. I looked in the spark log referenced by the spark context and I see this:
Caused by: java.net.URISyntaxException: Relative path in absolute URI: file:C:/Users/Stephen%20Upton/Documents/spark-warehouse
	at java.net.URI.checkPath(URI.java:1823)
	at java.net.URI.<init>(URI.java:745)
	at org.apache.hadoop.fs.Path.initialize(Path.java:202)

I can't seem to figure out why it sparklyr is not setting the spark-warehouse config variable. I'm hoping this is easy to fix/find.

thanx
steve

**Comment by yarrick19** [comment]
@menichols74 
We're you able to resolve your issue?  If so, how?  I appear to be having the same issue as you.

**Comment by kevinushey** [comment]
I wonder if the issue here is caused by spaces in the user name? @yarrick19, do you have any spaces or non-ASCII characters in your user name?

**Comment by yarrick19** [comment]
@kevinushey 
Moved your question and my response over to #277.

**Comment by javierluraschi** [comment]
@upton9265 could you jump to this issue https://github.com/rstudio/sparklyr/issues/277? We are troubleshooting there the `NativeConstructorAccessorImpl.newInstance0` exception for Windows systems.

**Comment by prbhadra** [comment]
Hi,
I have installed Spark 2.1.0 on windows 10 and configured the Winutils, HADOOP_HOME, SPARK_HOME as advised.

I am facing problem connecting the spark instance. Please find the log below and suggest:

version
_
platform x86_64-w64-mingw32
arch x86_64
os mingw32
system x86_64, mingw32
status
major 3
minor 3.2
year 2016
month 10
day 31
svn rev 71607
language R
version.string R version 3.3.2 (2016-10-31)
nickname Sincere Pumpkin Patch
library(sparklyr)
spark_install(version = "2.1.0")
Spark 2.1.0 for Hadoop 2.7 or later already installed.
sc <- spark_connect(master = "local")
Error in file(con, "r") : cannot open the connection
In addition: Warning messages:
1: In prepare_windows_environment(spark_home) :
HADOOP_HOME was already but does not match current Spark installation
2: In file(con, "r") :
cannot open file 'C:\Users\prbha\AppData\Local\Temp\Rtmpe6M86n\file2ac810972c5d_spark.log': Permission denied
sc <- spark_connect(master = "spark://192.168.199.1:7077")
Error in file(con, "r") : cannot open the connection
In addition: Warning messages:
1: In prepare_windows_environment(spark_home) :
HADOOP_HOME was already but does not match current Spark installation
2: In file(con, "r") :
cannot open file 'C:\Users\prbha\AppData\Local\Temp\Rtmpe6M86n\file2ac8689b538c_spark.log': Permission denied

**Comment by JKcme** [comment]
I've gotten the same problem.  All packages/versions are up to date.   Any advice?  I just went through the new Spark tab on RStudio 1.0.136.  I'm a total noob to this Spark stuff....thx!

The exact error:

Error in start_shell(master = master, spark_home = spark_home, spark_version = version,  : 
  Failed to launch Spark shell. Ports file does not exist.
    Path: C:\Users\E19665\AppData\Local\rstudio\spark\Cache\spark-1.6.2-bin-hadoop2.6\bin\spark-submit.cmd
    Parameters: --class, sparklyr.Backend, --packages, "com.databricks:spark-csv_2.11:1.3.0","com.amazonaws:aws-java-sdk-pom:1.10.34", "C:\Users\E19665\Documents\R\win-library\3.3\sparklyr\java\sparklyr-1.6-2.10.jar", C:\Users\E19665\AppData\Local\Temp\Rtmpe8bQdQ\file293031055d20.out
    
Error occurred during initialization of VM
Initial heap size set to a larger value than the maximum heap sizePicked up _JAVA_OPTIONS: -Xmx512M
Picked up _JAVA_OPTIONS: -Xmx512M 

My sessionInfo():

R version 3.3.2 (2016-10-31)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 7 x64 (build 7601) Service Pack 1

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                           LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] dplyr_0.5.0          sparklyr_0.4         RODBCDBI_0.1.1       RDCOMClient_0.93-0.2 RevoUtilsMath_10.0.0

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.8      withr_1.0.2      digest_0.6.11    rprojroot_1.1    assertthat_0.1   rappdirs_0.3.1   R6_2.2.0         DBI_0.5-1        backports_1.0.4 
[10] magrittr_1.5     httr_1.2.1       devtools_1.12.0  config_0.2       splines_3.3.2    RevoUtils_10.0.2 RODBC_1.3-14     tools_3.3.2      parallel_3.3.2  
[19] yaml_2.1.14      memoise_1.0.0    tibble_1.2

**Comment by javierluraschi** [comment]
Fix commited, please retry with `sparklyr 0.5.6-9002` or newer. You can install using `devtools::install_github("rstudio/sparklyr")`.

**Comment by bsr7890** [comment]
I was facing the same Gateway issue as mentioned in @yarrick19's post.

My Scenario:
Machine: Windows 7, 64bit

I wanted to work on R Studio, with spark in background.
Below are steps:

## THIS STEP IS VERY IMPORTANT for 64 bit machines.
Sys.setenv(JAVA_HOME='C:\\Program Files\\Java\\jre7') # for 64-bit version

#### For 32 bit machines: Sys.setenv(JAVA_HOME="C:\\Program Files\\Java\\jdk1.7.0_67\\bin")

devtools::install_github("rstudio/sparklyr")
install.packages("dplyr")
install.packages("DBI")
install.packages("Rcpp")

library(Rcpp)
library(DBI)
library(dplyr)
library(sparklyr)

spark_install(version = "2.0.0", hadoop_version = "2.7")

config <- spark_config()
config$sparklyr.gateway.port = 10000
config$sparklyr.gateway.connect.timeout = 1

sc <- spark_connect(master="local" #"yarn-client"
                    , method = c("shell")
                    ,config=config
                    , app_name = "sparklyr"
                    ,version="2.0.0"
                    , extensions = sparklyr::registered_extensions()
                    , spark_home = normalizePath(path.expand(path = "C:/Users/<USERNAME>/AppData/Local/rstudio/spark/Cache/spark-2.0.0-bin-hadoop2.7"), winslash = "\\", mustWork = NA)
                    )

After following above steps, I was able to make the cluster up in RStudio.

Thanks.
